Use steepest descent without thermostat in tutorial 02

jngrad opened the pull request:
Fixes #3030

Description of changes:
- cleanup and document the steepest descent algorithm
- reflect recent changes to the Python interface in the tutorials: `mindist()` renamed to `min_dist()`, periodicity takes a list of booleans
- in tutorial 02 charged system:
    - set Langevin thermostat after energy minimization
    - replace adaptive force-cap minimization strategy with steepest descent

jngrad committed:
Refactor minimization

Cleanup and remove unused bookkeeping of max displacement.

jngrad committed:
Move struct to header file

jngrad committed:
Document steepest descent algorithm

jngrad committed:
Update documentation with recent syntax changes

system.analysis.mindist() renamed to system.analysis.min_dist()
and system.periodicity now takes an array of bool

jngrad committed:
New energy minimization strategy

Replace the adaptive force cap minimization strategy with steepest
descent and enable the Langevin thermostat only after minimization.

jngrad commented:
Notes:
- part 1 of the tutorial still stops minimization based on the minimal distance criterion, we can remove that and have a pure steepest descent instead, just like part 2 of the tutorial
- some NaCl scripts use steepest descent with thermostat ([`nacl_units_confined.py`](https://github.com/espressomd/espresso/blob/python/doc/tutorials/02-charged_system/scripts/nacl_units_confined.py), [`nacl_units_confined_vis.py`](https://github.com/espressomd/espresso/blob/python/doc/tutorials/02-charged_system/scripts/nacl_units_confined_vis.py)), or no minimization at all ([`nacl_units.py`](https://github.com/espressomd/espresso/blob/python/doc/tutorials/02-charged_system/scripts/nacl_units.py), [`nacl_units_vis.py`](https://github.com/espressomd/espresso/blob/python/doc/tutorials/02-charged_system/scripts/nacl_units_vis.py))

fweik commented on src/core/minimize_energy.cpp:
```diff
@@ -81,11 +69,11 @@ bool steepest_descent_step(const ParticleRange &particles) {
           // Positional increment
-          dp = params->gamma * p.f.f[j];
-          if (fabs(dp) > params->max_displacement)
+          auto dp = params->gamma * p.f.f[j];
```
You can use `clamp` to allow for it being const:
```c++
auto const dp = boost::clamp(gamma * f, -max_displacement, max_displacement);
```

KaiSzuttor commented on src/core/minimize_energy.cpp:
```diff
@@ -40,32 +40,20 @@
 
-/* Signum of val */
+/** Sign of the argument */
 template <typename T> int sgn(T val) { return (T(0) < val) - (val < T(0)); }
```
```src/utils/include/utils/math/sgn.hpp```

jngrad committed:
Use boost::algorithm::clamp

jngrad committed:
Minimization with custom convergence criterion

When running minimization in a loop, if the force on particles is
lower than f_max, the minimization will run only one integration
step when minimize() is called, slowing down convergence by a
factor of max_steps. Setting f_max to 0 prevents that.

fweik commented:
bors r+

jngrad commented:
The CI pipeline apparently canceled itself because the merging happened before the staging: [943f3e2c/staging](https://gitlab.icp.uni-stuttgart.de/espressomd/espresso/commit/943f3e2cc28f4c866579729728cae11449932a2b) then [73b71c12/staging-tmp](https://gitlab.icp.uni-stuttgart.de/espressomd/espresso/commit/73b71c12f5612d6c03304ad3929989aafa41992b).

jngrad commented:
bors r-

jngrad commented:
bors retry

jngrad commented:
bors retry

jngrad commented:
bors r=fweik

KaiSzuttor commented:
bros

KaiSzuttor commented:
bors retry
